### PR TITLE
 - APEXMALHAR-2070 #resolve #comment

### PR DESCRIPTION
**Pull request by ilooner**
- Added SpillableMultiSetInterface
- Added in memory implementations of SpillableArrayList, SpillableMultiset, and SpillableByteArrayListMultimap

### CODE REVIEW

File: library/src/main/java/org/apache/apex/malhar/lib/state/spillable/SpillableComplexComponent.java [hunk 1]
```diff
@@ -105,6 +105,26 @@
...
+
+  /**
+   * This is a method for creating a {@link SpillableByteMultiset}.
+   * @param <T> The type of the elements.
+   * @param identifier The identifier for this {@link SpillableByteMultiset}.
+   * @param bucket The bucket that this {@link SpillableByteMultiset} will be spilled too.
+   * @param serde The Serializer/Deserializer to use for data stored in the {@link SpillableByteMultiset}.
+   * @return A {@link SpillableByteMultiset}.
+   */
+  <T> SpillableByteMultiset<T> newSpillableMultiset(byte[] identifier, long bucket, Serde<T, byte[]> serde);
```
Comment (tweise) [inline library/src/main/java/org/apache/apex/malhar/lib/state/spillable/SpillableComplexComponent.java]: How will this work with ManagedState? Will the data be distributed over multiple data files, partitioned by time?

Comment (ilooner) [inline library/src/main/java/org/apache/apex/malhar/lib/state/spillable/SpillableComplexComponent.java]: @tweise The multiset just keeps track of each element and its count. That could be represented as a map.  Event time could be extracted from the tuple or processing time could be used to distribute the data over multiple data files.

**Comment by ilooner** [comment]
@chandnisingh @tweise please review

File: library/src/main/java/org/apache/apex/malhar/lib/state/spillable/Spillable.java [hunk 2]
```diff
@@ -44,8 +45,10 @@
   }
 
   /**
-   * This represents a spillable {@link java.util.Map}. The underlying implementation
-   * of this map uses the serialized representation of the key object as a key. User's that receive an
+   * This represents a spillable {@link java.util.Map}. The implementations assume
+   * that the fields from objects of type K that are serialized are the same fields that are used in the
```
Comment (otterc) [inline library/src/main/java/org/apache/apex/malhar/lib/state/spillable/Spillable.java]: I think the comment can be simplified using K1 and K2. The assumption is that k1.equals(k2) should be consistent with  k1.toByteArray().equals(k2.toByteArray())

Comment (ilooner) [inline library/src/main/java/org/apache/apex/malhar/lib/state/spillable/Spillable.java]: @chandnisingh fixed documentation as you suggested.

**Comment by ilooner** [comment]
@chandnisingh updated interfaces to use Slice as suggested by Thomas

File: library/src/main/java/org/apache/apex/malhar/lib/state/spillable/inmem/InMemMultiset.java [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+
+import javax.annotation.Nullable;
+
+import org.apache.apex.malhar.lib.state.spillable.Spillable;
+
+import com.esotericsoftware.kryo.serializers.FieldSerializer;
+import com.esotericsoftware.kryo.serializers.JavaSerializer;
+import com.google.common.collect.HashMultiset;
+
+public class InMemMultiset<T> implements Spillable.SpillableByteMultiset<T>
```
Comment (otterc) [inline library/src/main/java/org/apache/apex/malhar/lib/state/spillable/inmem/InMemMultiset.java:37]: Java doc missing

File: library/src/main/java/org/apache/apex/malhar/lib/state/spillable/inmem/InMemSpillableArrayList.java [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.List;
+import java.util.ListIterator;
+
+import org.apache.apex.malhar.lib.state.spillable.Spillable;
+
+import com.google.common.collect.Lists;
+
+public class InMemSpillableArrayList<T> implements Spillable.SpillableArrayList<T>
```
Comment (otterc) [inline library/src/main/java/org/apache/apex/malhar/lib/state/spillable/inmem/InMemSpillableArrayList.java:34]: Javadoc missing

File: library/src/main/java/org/apache/apex/malhar/lib/state/spillable/inmem/InMemSpillableByteArrayListMultimap.java [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+import org.apache.apex.malhar.lib.state.spillable.Spillable;
+
+import com.esotericsoftware.kryo.serializers.FieldSerializer;
+import com.esotericsoftware.kryo.serializers.JavaSerializer;
+import com.google.common.collect.ArrayListMultimap;
+import com.google.common.collect.ListMultimap;
+import com.google.common.collect.Multimap;
+import com.google.common.collect.Multiset;
+
+public class InMemSpillableByteArrayListMultimap<K, V> implements Spillable.SpillableByteArrayListMultimap<K, V>
```
Comment (otterc) [inline library/src/main/java/org/apache/apex/malhar/lib/state/spillable/inmem/InMemSpillableByteArrayListMultimap.java:42]: Javadoc missing

File: library/src/main/java/org/apache/apex/malhar/lib/state/spillable/inmem/InMemSpillableComplexComponent.java [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+ */
+package org.apache.apex.malhar.lib.state.spillable.inmem;
+
+import org.apache.apex.malhar.lib.state.spillable.SpillableComplexComponent;
+import org.apache.apex.malhar.lib.utils.serde.Serde;
+
+import com.datatorrent.api.Context;
+import com.datatorrent.netlet.util.Slice;
+
+public class InMemSpillableComplexComponent implements SpillableComplexComponent
```
Comment (otterc) [inline library/src/main/java/org/apache/apex/malhar/lib/state/spillable/inmem/InMemSpillableComplexComponent.java:30]: Javadoc missing

File: library/src/test/java/org/apache/apex/malhar/lib/state/managed/spillable/inmem/InMemMultisetTest.java [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+import org.junit.Assert;
+import org.junit.Test;
+
+import org.apache.apex.malhar.lib.state.spillable.inmem.InMemMultiset;
+
+import com.esotericsoftware.kryo.Kryo;
+
+import com.datatorrent.lib.util.KryoCloneUtils;
+
+public class InMemMultisetTest
```
Comment (otterc) [inline library/src/test/java/org/apache/apex/malhar/lib/state/managed/spillable/inmem/InMemMultisetTest.java:30]: Javadoc missing

**Comment by otterc** [comment]
@ilooner please add the javadocs to new classes added and then I will merge it.

Thanks,
Chandni

File: library/src/main/java/org/apache/apex/malhar/lib/state/spillable/inmem/InMemMultiset.java (hunk 3 above)
Comment (ilooner) [inline library/src/main/java/org/apache/apex/malhar/lib/state/spillable/inmem/InMemMultiset.java:37]: Done

File: library/src/main/java/org/apache/apex/malhar/lib/state/spillable/inmem/InMemSpillableArrayList.java (hunk 4 above)
Comment (ilooner) [inline library/src/main/java/org/apache/apex/malhar/lib/state/spillable/inmem/InMemSpillableArrayList.java:34]: Done

File: library/src/main/java/org/apache/apex/malhar/lib/state/spillable/inmem/InMemSpillableByteArrayListMultimap.java (hunk 5 above)
Comment (ilooner) [inline library/src/main/java/org/apache/apex/malhar/lib/state/spillable/inmem/InMemSpillableByteArrayListMultimap.java:42]: Done

File: library/src/test/java/org/apache/apex/malhar/lib/state/managed/spillable/inmem/InMemMultisetTest.java (hunk 7 above)
Comment (ilooner) [inline library/src/test/java/org/apache/apex/malhar/lib/state/managed/spillable/inmem/InMemMultisetTest.java:30]: This is a test class

**Comment by ilooner** [comment]
@chandnisingh added javadoc

**Pull request merged**